### PR TITLE
DrivAerML: gc=0.5+T_max=10 — softer clip, faster cycling

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

DrivAerML golden config uses T_max=30 and no gradient clipping. AirfRANS's breakthrough came from combining gc=1.0 with tight T_max=5. For DrivAerML, gc=1.0 is being tested by faye and franky with T_max=30. This PR explores the orthogonal axis: a softer gradient clip (gc=0.5) with faster cosine cycling (T_max=10). The rationale is that gc=0.5 is more permissive than gc=1.0 but still regularizes large gradient steps, while T_max=10 explores a middle ground between DrivAerML's T_max=30 and AirfRANS's T_max=5. We also extend to AirfRANS gc=0.5 variants for cross-dataset comparison.

## Instructions

Run all 8 commands in parallel, one per GPU. All commands run from `cd target/icml2026`. Do NOT set SENPAI_TIMEOUT_MINUTES or SENPAI_MAX_EPOCHS — inherit from pod env.

**GPU 0 — DrivAerML: lr=5e-4, gc=0.5, T_max=10**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 10 \
  --grad-clip 0.5 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --wandb-group gc05-tmax10
```

**GPU 1 — DrivAerML: lr=5e-4, gc=0.5, T_max=20**
```bash
CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 20 \
  --grad-clip 0.5 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --wandb-group gc05-tmax10
```

**GPU 2 — DrivAerML: lr=6e-4, gc=0.5, T_max=10**
```bash
CUDA_VISIBLE_DEVICES=2 python train.py \
  --dataset drivaerml --optimizer adamw --lr 6e-4 --cosine-t-max 10 \
  --grad-clip 0.5 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --wandb-group gc05-tmax10
```

**GPU 3 — DrivAerML: lr=7e-4, gc=0.5, T_max=10**
```bash
CUDA_VISIBLE_DEVICES=3 python train.py \
  --dataset drivaerml --optimizer adamw --lr 7e-4 --cosine-t-max 10 \
  --grad-clip 0.5 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --wandb-group gc05-tmax10
```

**GPU 4 — DrivAerML: lr=5e-4, gc=0.5, T_max=10, wd=1e-2**
```bash
CUDA_VISIBLE_DEVICES=4 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 10 \
  --grad-clip 0.5 --weight-decay 1e-2 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --wandb-group gc05-tmax10
```

**GPU 5 — DrivAerML: lr=5e-4, gc=2.0, T_max=10 (looser clip)**
```bash
CUDA_VISIBLE_DEVICES=5 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 10 \
  --grad-clip 2.0 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --wandb-group gc05-tmax10
```

**GPU 6 — AirfRANS: lr=7e-4, gc=0.5, T_max=5, wd=1e-2 (softer clip vs golden)**
```bash
CUDA_VISIBLE_DEVICES=6 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 0.5 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 \
  --model-heads 4 --epochs 999 --wandb-group gc05-tmax10
```

**GPU 7 — AirfRANS: lr=7e-4, gc=2.0, T_max=5, wd=1e-2 (looser clip vs golden)**
```bash
CUDA_VISIBLE_DEVICES=7 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 2.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 \
  --model-heads 4 --epochs 999 --wandb-group gc05-tmax10
```

## Baseline

**DrivAerML** (primary target for GPUs 0–5):
- `val_primary/surface_rel_l2_pct = 4.619%` — PR #2691 (frieren), epoch 256/267
- Architecture: 4L/512d/8H, AdamW lr=5e-4, T_max=30, no gc, no wd
- External target: <3.71%

**AirfRANS** (GPUs 6–7):
- `val_primary/surface_mse = 0.001236` — PR #2828 (inosuke), epoch 349/388
- Architecture: 2L/256d/4H, AdamW lr=7e-4, gc=1.0, wd=1e-2, T_max=5

**TandemFoil** (not tested this PR):
- `val_primary/surface_pressure_mae = 30.10` — PR #2810 (gilbert), epoch 157

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT gojo:

## Results

### DrivAerML (GPUs 0–5) — Baseline: `val_primary/surface_rel_l2_pct = 4.619%` (external target <3.71%)

| GPU | Config | Best Val % | Test % | Epochs | Grad Norm | Status |
|-----|--------|-----------|--------|--------|-----------|--------|
| 0 | lr=5e-4, gc=0.5, T_max=10 | ~70.5 | 68.72 | 367 | inf | **FAIL** |
| 1 | lr=5e-4, gc=0.5, T_max=20 | **4.78** | 6.19 | 371 | 0.178 | Worse than baseline |
| 2 | lr=6e-4, gc=0.5, T_max=10 | ~64.6 | 66.75 | 375 | inf | **FAIL** |
| 3 | lr=7e-4, gc=0.5, T_max=10 | ~84.9 | 83.29 | 357 | inf | **FAIL** |
| 4 | lr=5e-4, gc=0.5, T_max=10, wd=1e-2 | ~79.0 | 81.77 | 358 | inf | **FAIL** |
| 5 | lr=5e-4, gc=2.0, T_max=10 | ~43.9 | 44.62 | 366 | 9e16 | **FAIL** |

### AirfRANS (GPUs 6–7) — Baseline: `val_primary/surface_mse = 0.001236`

| GPU | Config | Best Val MSE | Test MSE | Epochs | Grad Clips/Batch | Status |
|-----|--------|-------------|----------|--------|-----------------|--------|
| 6 | lr=7e-4, gc=0.5, T_max=5, wd=1e-2 | 0.001501 | 0.001548 | 678 | ~210/360 (58%) | Worse than baseline |
| 7 | lr=7e-4, gc=2.0, T_max=5, wd=1e-2 | **0.001065** | 0.006872 | 676 | ~60/360 (17%) | Val beats baseline but test regressed badly |

### Peak Memory
- DrivAerML: ~6 GB per GPU (observed via nvidia-smi)
- AirfRANS: ~25–33 GB per GPU (observed via nvidia-smi)

### Commands Used
```bash
# GPU 0–5: DrivAerML variants (shown is GPU 0 template)
CUDA_VISIBLE_DEVICES=0 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 10 --grad-clip 0.5 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group gc05-tmax10 --agent gojo --wandb-name "gojo/drivaer-lr5e4-gc0.5-tmax10"

# GPU 6: AirfRANS gc=0.5
CUDA_VISIBLE_DEVICES=6 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 0.5 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --wandb-group gc05-tmax10 --agent gojo --wandb-name "gojo/airfrans-lr7e4-gc0.5-tmax5"

# GPU 7: AirfRANS gc=2.0
CUDA_VISIBLE_DEVICES=7 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 2.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --wandb-group gc05-tmax10 --agent gojo --wandb-name "gojo/airfrans-lr7e4-gc2.0-tmax5"
```

### W&B Run IDs
- GPU 0: [6v1t3w6f](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/6v1t3w6f)
- GPU 1: [zx6ndjjo](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/zx6ndjjo)
- GPU 2: [alfbc1vu](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/alfbc1vu)
- GPU 3: [xla6tsgo](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/xla6tsgo)
- GPU 4: [9u1t40el](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/9u1t40el)
- GPU 5: [pi9rt898](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/pi9rt898)
- GPU 6: [fyn4ni5r](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/fyn4ni5r)
- GPU 7: [cz18ip2p](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/cz18ip2p)

### What happened

**DrivAerML T_max=10 is catastrophically unstable.** All five T_max=10 variants (GPUs 0, 2, 3, 4, 5) produced Infinity or astronomical (~9e16) grad norms with every single batch being clipped. The models essentially never learned — val errors ranged from 43% to 85% vs the 4.619% baseline. Even gc=2.0 (a much looser clip) couldn't rescue T_max=10. This strongly suggests T_max=10 causes the cosine schedule to reset too aggressively for DrivAerML's 394-batch training loop, creating destructive learning rate oscillations that gradient clipping alone cannot stabilize.

**T_max=20 (GPU 1) was the sole survivor** at 4.78% val, but still ~3.5% worse than the 4.619% baseline. The healthy grad norm (0.178) and minimal clipping (6/394 batches) confirms that slower cycling is critical. DrivAerML clearly needs longer cosine periods than AirfRANS — the current baseline's T_max=30 remains superior.

**AirfRANS gc=0.5 (GPU 6) underperformed gc=1.0 baseline.** Best val 0.001501 vs baseline 0.001236 (+21.4%). The softer clip caused 58% of batches to be clipped (vs the golden gc=1.0 which clips fewer). This suggests gc=0.5 is too permissive for AirfRANS — it doesn't provide enough gradient regularization.

**AirfRANS gc=2.0 (GPU 7) shows interesting but unreliable behavior.** Best val 0.001065 would beat baseline by 13.8%, but test MSE was 0.006872 (5.6x worse than val). This massive val/test gap indicates severe overfitting — gc=2.0 is too loose, allowing the model to overfit to validation without generalizing. The golden gc=1.0 clearly hits the right balance.

### Suggested follow-ups

1. **DrivAerML T_max sweep at larger values**: T_max=25 might find a sweet spot between the baseline T_max=30 and the failed T_max=10/20. The data shows DrivAerML needs slower cycling than AirfRANS.
2. **DrivAerML gc=1.0 + T_max=30**: Since gc=0.5 and gc=2.0 failed at T_max=10, the AirfRANS golden recipe of gc=1.0 might still help at DrivAerML's native T_max=30 (being tested by faye/franky).
3. **AirfRANS gc=1.0 remains the sweet spot**: gc=0.5 under-regularizes (too many clips), gc=2.0 over-fits. The current golden gc=1.0 is well-calibrated for AirfRANS.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2888 because ADVISOR: Closing — dead end. gc=0.5+T_max=10 had 5 of 6 DrivAerML runs diverge fatally (grad norms → inf). Only the T_max=20 variant survived at 4.78% — worse than baseline. AirfRANS also regressed (0.001501). T_max=10 + gc=0.5 is incompatible with DrivAerML..
